src/git.rs

```diff
@@ -118,7 +118,61 @@ pub enum Repository {
 }
 
 impl Repository {
-    pub fn git(url: url::Url) -> Self {
+    /// Takes in only an URL and tries to find the most specific repository type that matches. Falls back to `Self::Git`
+    ///
+    /// For forges that can also be self-hosted, only the well-known "main" forge will be taken into account (e.g. gitlab.com).
+    /// Only works for HTTP(S) URLs.
+    /// Invalid URLs will equally fall back to `Self::Git`, as we can't know for sure what is invalid without trying it out later.
+    pub async fn git_auto(url: Url) -> Self {
+        let url2 = url.clone();
+        match (url.scheme(), url.domain()) {
+            ("http" | "https", Some("github.com")) => Self::github_from_url(url),
+            ("http" | "https", Some("gitlab.com")) => Self::gitlab_from_url(url),
+            ("http" | "https", Some("forgejo.org")) => Self::forgejo_from_url(url),
+            ("http" | "https", _) => Self::probe_forge(url).await,
+            _ => None,
+        }
+        .unwrap_or(Self::git(url2))
+    }
+
+    ///
+    /// Takes in a URL of unknown forge and tries to determine which forge the hoster is
+    /// And then parse the url into the according Repository Variant
+    async fn probe_forge(url: Url) -> Option<Self> {
+        async fn probe(mut test_url: Url, path: &str) -> Result<bool> {
+            test_url.set_path(path);
+            let res = reqwest::get(test_url).await?;
+            Ok(res.status() != 404)
```

Comment:
Given that both APIs will return JSON, I'd argue that the response should successfully parse as such. In its current form, I'd be worried that some server throwing a 500 or 403 might yield a false positive